adapter/src/test/kotlin/org/javacs/ktda/GlobalWorkspaceTest.kt

```diff
@@ -0,0 +1,242 @@
+import org.javacs.ktda.jdi.launch.JDILauncher
+import org.junit.After
+import org.junit.Assert.assertThat
+import org.junit.Before
+import org.hamcrest.Matchers.equalTo
+import org.junit.Test
+import java.lang.Thread
+import java.util.concurrent.CountDownLatch
+
+/**
+ * Defines the basic debugging scenarios
+ * that should be tested on all supported sample applications.
+ */
+abstract class GlobalWorkspaceTest(
+    relativeWorkspaceRoot: String,
+    private val mainClass: String,
+    private val defaultVmArguments: String = "",
+    private val buildCommand: Pair<String, String>
+) {
+    val absoluteWorkspaceRoot: Path =
+        Paths.get(GlobalWorkspaceTest::class.java.getResource("/Anchor.txt").toURI()).parent.resolve(
+            relativeWorkspaceRoot
+        )
+    var client = mockk<IDebugProtocolClient>(relaxed = true)
+    lateinit var debugAdapter: KotlinDebugAdapter
+    val latch = CountDownLatch(1)
+
+    @Before
+    fun startDebugAdapter() {
+        // Reset the mock debug protocol client
+        client = mockk(relaxed = true)
+
+        // Build the project first
+        val process = ProcessBuilder(buildCommand.first, buildCommand.second)
+            .directory(absoluteWorkspaceRoot.toFile())
+            .inheritIO()
+            .start()
+        process.waitFor()
+        assertThat(process.exitValue(), equalTo(0))
+
+        debugAdapter = JDILauncher()
+            .let(::KotlinDebugAdapter)
+            .also {
+                it.connect(client)
+                val configDone = it.configurationDone(ConfigurationDoneArguments())
+                it.initialize(InitializeRequestArguments().apply {
+                    adapterID = "test-debug-adapter"
+                    linesStartAt1 = true
+                    columnsStartAt1 = true
+                }).join()
+                // Slightly hacky workaround to ensure someone is
+                // waiting on the ConfigurationDoneResponse. See
+                // KotlinDebugAdapter.kt:performInitialization for
+                // details.
+                Thread {
+                    configDone.join()
+                }.start()
+                // Wait until the thread has blocked on the future
+                while (configDone.numberOfDependents == 0) {
+                    Thread.sleep(100)
+                }
+            }
+
+        // Print output on commandline for test debugging
+        slot<OutputEventArguments>().also { outputEvent ->
+            every { client.output(capture(outputEvent)) }.answers { println(outputEvent.captured.output) }
+        }
+        // And release the latch when the debug session is terminated/Done
+        every { client.terminated(any()) }.answers { latch.countDown() }
+    }
+
+    @After
+    fun closeDebugAdapter() {
+        debugAdapter.disconnect(DisconnectArguments()).join()
+    }
+
+    @Test
+    fun `Breakpoints should be hit and variables recorded`() {
+        // Given: We set a breakpoint
```

Comment:
Feel free to hate the Given When Then style of test comments.
Its what I'm used to I wanted to suggest it. 100% not a hill I will die on if you prefer otherwise.